### PR TITLE
Feature/associations

### PR DESCRIPTION
**Pull request by skyv26**
# Milestone #3 - Set up the associations

> In this feature branch, I added the `associations`, and `relationship`, between the multiple classes as per the diagram into files and implemented the features for creating the books, rentals, students, and teachers. Details of the features that I have implemented in this branch are as follows:

- Created different folders for different files.
- Modified the `Person` class by inheriting the features of the *Nameable* class interface.
- Added the feature by using the association's concept.
- Implemented the **function overriding**  `feature` to extend the functionality of the and also added the `self` for passing the class as an object. 
- Resolved the `rubocop` linter errors.
- Used the *best practices* for writing the code in `Ruby`.
- Converted all the variables and method names in `SnakeCase` as per Ruby Styling Guide.

### CODE REVIEW

File: lib/student.rb [hunk 1]
```diff
@@ -9,4 +9,9 @@ def initialize(age, classroom, name = 'Unknown', parent_permission: true)
   def play_hooky
     '¯\(ツ)/¯'
   end
+
+  def classroom=(classroom)
+    @classroom = classroom
```
Comment (FrederickMih) [inline lib/student.rb:16]: - [x] You didn't declare the classroom setters instance variable in this class.

```ruby
class Student < Person
   attr_reader :classroom
    .
    .

```

Let's kindly fix :wrench: it

File: lib/book/book.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class Book
+  attr_accessor :title, :author
```
Comment (FrederickMih) [inline lib/book/book.rb]: - [x] We equally need Getters and Setters for the @rentals instance variable to complete the `A book has many rentals` relationship. Kindly fix it. `attr_accessor :title, :author, :rentals`

File: lib/book/book.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  def initialize(title = 'Unknown', author = 'Unknown')
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add(rental)
+    rental.book = self
+    @rentals << rental
+  end
```
Comment (FrederickMih) [inline lib/book/book.rb]: - [**Optional**] We could also do it like this:

```ruby
    def add(person, date)
       Rental.new(date, self, person)
    end
```

File: lib/person.rb [hunk 4]
```diff
@@ -11,6 +11,12 @@ def initialize(age, name = 'Unknown', parent_permission: true)
     @name = name
     @age = age
     @parent_permission = parent_permission
+    @rentals = []
+  end
+
+  def add_rental(rental)
```
Comment (FrederickMih) [inline lib/person.rb]: - [x] Getters and Setters for @rentals should be present

  ```ruby
  class Person
    

       attr_accessor :name, :age, :rentals
```

File: lib/person.rb [hunk 5]
```diff
@@ -11,6 +11,12 @@ def initialize(age, name = 'Unknown', parent_permission: true)
...
+    @rental << rental
+    rental.person = self
   end
```
Comment (FrederickMih) [inline lib/person.rb]: -  You may kindly refer to the comment above and fix this method.

**Review by FrederickMih — CHANGES_REQUESTED** [review]
Hi @skyv26, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


Screenshot from 2022-10-20 09-38-51


## Highlights :green_circle: 


- Correct Gitflow has been followed :heavy_check_mark: 
- Good commit history :heavy_check_mark: 
- PR has a good title and summary :heavy_check_mark: 
- No Linter error :heavy_check_mark: 


## Required Changes ♻️

N/A


_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: lib/student.rb (hunk 1 above)
Comment (skyv26) [reply lib/student.rb:16]: Thanks fixed

File: lib/book/book.rb (hunk 2 above)
Comment (skyv26) [reply lib/book/book.rb]: Fixed

File: lib/book/book.rb (hunk 3 above)
Comment (skyv26) [reply lib/book/book.rb]: Done

File: lib/person.rb (hunk 4 above)
Comment (skyv26) [reply lib/person.rb]: Fixed

File: lib/person.rb (hunk 5 above)
Comment (skyv26) [reply lib/person.rb]: Thanks fixed all.

**Review by Wes-Isaac — APPROVED** [review]
Hi @skyv26, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- Professional README 👍.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by skyv26** [comment]
@Wes-Isaac Thank you very much ❤️ 

Happy Coding 🥂

**Pull request merged**